src/person.cpp

```diff
@@ -1711,40 +1707,21 @@ void Person::DecideMoveDirectionOnPathlessLand(Voxel *former_voxel,
 	auto convert_slope = [](TileSlope &slope) {
 		if ((slope & TSB_STEEP) == 0 || (slope & TSB_TOP) != 0) return;
 		const TileSlope raised_corner_bit = (slope & ~TSB_STEEP);
-		slope = TSB_MASK_ALL ^ raised_corner_bit;
+		if (raised_corner_bit != TSB_NORTH) slope |= TSB_SOUTH;
+		if (raised_corner_bit != TSB_SOUTH) slope |= TSB_NORTH;
+		if (raised_corner_bit != TSB_WEST ) slope |= TSB_EAST;
+		if (raised_corner_bit != TSB_EAST ) slope |= TSB_WEST;
 	};
 	convert_slope(old_voxel_slope);
 	convert_slope(new_voxel_slope);
 
-	uint16 left_height_1, left_height_2, right_height_1, right_height_2;
-	switch (exit_edge) {
-		case EDGE_NE:
-			left_height_1  = former_vox_pos.z + ((old_voxel_slope & TSB_NORTH) == 0 ? 0 : 1);
-			right_height_1 = former_vox_pos.z + ((old_voxel_slope & TSB_EAST ) == 0 ? 0 : 1);
-			left_height_2  =  this->vox_pos.z + ((new_voxel_slope & TSB_WEST ) == 0 ? 0 : 1);
-			right_height_2 =  this->vox_pos.z + ((new_voxel_slope & TSB_SOUTH) == 0 ? 0 : 1);
-			break;
-		case EDGE_SE:
-			left_height_1  = former_vox_pos.z + ((old_voxel_slope & TSB_EAST ) == 0 ? 0 : 1);
-			right_height_1 = former_vox_pos.z + ((old_voxel_slope & TSB_SOUTH) == 0 ? 0 : 1);
-			left_height_2  =  this->vox_pos.z + ((new_voxel_slope & TSB_NORTH) == 0 ? 0 : 1);
-			right_height_2 =  this->vox_pos.z + ((new_voxel_slope & TSB_WEST ) == 0 ? 0 : 1);
-			break;
-		case EDGE_SW:
-			left_height_1  = former_vox_pos.z + ((old_voxel_slope & TSB_SOUTH) == 0 ? 0 : 1);
-			right_height_1 = former_vox_pos.z + ((old_voxel_slope & TSB_WEST ) == 0 ? 0 : 1);
-			left_height_2  =  this->vox_pos.z + ((new_voxel_slope & TSB_EAST ) == 0 ? 0 : 1);
-			right_height_2 =  this->vox_pos.z + ((new_voxel_slope & TSB_NORTH) == 0 ? 0 : 1);
-			break;
-		case EDGE_NW:
-			left_height_1  = former_vox_pos.z + ((old_voxel_slope & TSB_WEST ) == 0 ? 0 : 1);
-			right_height_1 = former_vox_pos.z + ((old_voxel_slope & TSB_NORTH) == 0 ? 0 : 1);
-			left_height_2  =  this->vox_pos.z + ((new_voxel_slope & TSB_SOUTH) == 0 ? 0 : 1);
-			right_height_2 =  this->vox_pos.z + ((new_voxel_slope & TSB_EAST ) == 0 ? 0 : 1);
-			break;
-		default: NOT_REACHED();
+	uint16 node_heights[4];
+	for (uint8 i = 0; i < 4; i++) {
+		const uint8 bitshift = (exit_edge + i) % 4;
+		node_heights[i] = (i < 2 ? former_vox_pos : this->vox_pos).z + (((i < 2 ? old_voxel_slope : new_voxel_slope) & (1 << bitshift)) >> bitshift);
```

Comment:
`(((i < 2 ? old_voxel_slope : new_voxel_slope) & (1 << bitshift)) >> bitshift)` Shifting 1 left, masking and then shifting right again can also be done by shifting right and then masking the lowest bit `(((i < 2 ? old_voxel_slope : new_voxel_slope) >> bitshift) & 1)`

Reply:
All implemented :)